repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AssetPreProcessor.java

```diff
@@ -0,0 +1,116 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+import static org.apache.atlas.repository.Constants.*;
+
+public class AssetPreProcessor implements PreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(AssetPreProcessor.class);
+
+    private EntityMutationContext context;
+    private AtlasTypeRegistry typeRegistry;
+    private EntityGraphRetriever entityRetriever;
+
+    public AssetPreProcessor(AtlasTypeRegistry typeRegistry, EntityGraphRetriever entityRetriever) {
+        this.typeRegistry = typeRegistry;
+        this.entityRetriever = entityRetriever;
+    }
+
+    @Override
+    public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context,
+                                  EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("AssetPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+        this.context = context;
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        switch (operation) {
+            case CREATE:
+                processCreateAsset(entity, vertex);
+                break;
+            case UPDATE:
+                processUpdateAsset(entity, vertex);
+                break;
+        }
+    }
+
+    private void processCreateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+          validateDomainAssetLinks(entity);
+          AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+          isAuthorized(sourceEntity);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+
+    private void processUpdateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+            validateDomainAssetLinks(entity);
+            AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+            isAuthorized(sourceEntity);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+
+    }
+
+    private void validateDomainAssetLinks(AtlasEntity entity) throws AtlasBaseException {
+        List<String> domainGuids = ( List<String>) entity.getAttribute(DOMAIN_GUIDS);
+
+        if(domainGuids.size() > 1) {
+            throw new AtlasBaseException(AtlasErrorCode.INVALID_PARAMETERS, "Asset can be linked to only one domain");
+        }
+
+        if(CollectionUtils.isNotEmpty(domainGuids)) {
+            for(String domainGuid : domainGuids) {
```

Comment:
since we already ensures size not greater than 1, no need to loop just process domainGuids.get(0)